File: java/javaNode/Interp.java
import java.io.*;
import java.util.*;

public class Interp extends Compiler {
	private Parser parser;
	private PrintWriter out;
	private StreamTokenizer in;
	private Stack<NodeInteger> stk;
	private Tabid<NodeInteger> vars;

	public static void main(String args[]) {
		Interp interp = new Interp(args[0]);
		if (interp.parse())
			interp.generate();
	}
	Interp(String infile) {
		try {
			parser = new Parser(new FileReader(infile));
		} catch (FileNotFoundException e) {
			System.err.println(infile + e);
			parser = new Parser(new InputStreamReader(System.in));
		}
		Writer outp = new PrintWriter(System.out);
		out = new PrintWriter(outp, true);
		Reader inp = new InputStreamReader(System.in);
		in = new StreamTokenizer(inp);
		in.eolIsSignificant(true);
		stk = new Stack<NodeInteger>();
		vars = new Tabid<NodeInteger>();
	}
	public boolean parse() { return parser.yyparse() == 0; }
	public boolean generate() {
		if (parser.errors() > 0) return false;
		evaluate(parser.syntax());
		return true;
	}
	public void evaluate(Node n) {
	  NodeList l;
	  String name;
	  int arg1, arg2;

	  switch (n.attrib()) {
	    case Parser.INTEGER: stk.push((NodeInteger)n); break;
	    case Parser.STRING: out.println(((NodeString)n).text()); break;
	  case Parser.STMT:
		l = (NodeList)n;
		for (int i = 0; i < l.size(); i++)
		  evaluate(l.get(i));
		break;
	  case Parser.VARIABLE: stk.push(vars.get(((NodeString)n).text())); break;
	  case Parser.WHILE:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		while (stk.pop().integer() != 0) {
			evaluate(l.second()); // eval instruction
			evaluate(l.first()); // eval condition, again...
		}
		break;
	  case Parser.IF:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		if (stk.pop().integer() != 0)
			evaluate(l.second()); // eval instruction
		break;
	  case Parser.ELSE:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		if (stk.pop().integer() != 0)
			evaluate(l.second()); // eval then
		else
			evaluate(l.third()); // eval else
		break;
	  case Parser.READ:
		l = (NodeList)n;
		name = ((NodeString)l.first()).text();
		if (vars.containsKey(name)) {
		    try {
			while (in.nextToken() != StreamTokenizer.TT_EOF) {
			    if (in.ttype == StreamTokenizer.TT_NUMBER) {
				vars.put(name, new NodeInteger(0,(int)in.nval));
				break;
			    }
			    if (in.ttype != StreamTokenizer.TT_EOL)
				System.err.println("Parse error at line "+in.lineno());
		        }
		    } catch (IOException e) {
		    	System.err.println("IO error: " + e);
			System.exit(2);
		    }
		}
		else
		    System.err.println(name + ": variable not found");
		break;
	  case Parser.PRINT:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		out.println(stk.pop().integer());
		break;
	  case Parser.ASSIGN:
		l = (NodeList)n;
		evaluate(l.second()); // eval expression to TOS
		vars.put(((NodeString)l.first()).text(), stk.pop());
		break;
	  case Parser.UMINUS:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		stk.push(new NodeInteger(0,-stk.pop().integer()));
		break;
	  case Parser.ADD:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 + arg2));
		break;
	  case Parser.SUB:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 - arg2));
		break;
	  case Parser.MUL:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 * arg2));
		break;
	  case Parser.DIV:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 / arg2));
		break;
	  case Parser.MOD:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 % arg2));
		break;
	  case Parser.EQ:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 == arg2 ? 1 : 0));
		break;
	  case Parser.NE:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 != arg2 ? 1 : 0));
		break;
	  case Parser.GE:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 >= arg2 ? 1 : 0));
		break;
	  case Parser.GT:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 > arg2 ? 1 : 0));
		break;
	  case Parser.LE:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 <= arg2 ? 1 : 0));
		break;
	  case Parser.LT:
		l = (NodeList)n;
		evaluate(l.first()); // eval condition to TOS
		arg1 = stk.pop().integer();
		evaluate(l.second()); // eval expression to TOS
		arg2 = stk.pop().integer();
		stk.push(new NodeInteger(0,arg1 < arg2 ? 1 : 0));
		break;
      }
    }
}

File: src/interp.c
/* $Id: interp.c,v 1.4 2012/09/15 01:14:18 prs Exp $ */
#include <stdio.h>
#include "node.h"
#include "y.tab.h"
#include "tabid.h"

static int ex(Node *p);
void evaluate(Node *p) { extern int errors; if (errors == 0) ex(p); }

static int ex(Node *p) {
  int i, val = 0;
  char *name;

  if (p == 0) return 0;
  switch(p->attrib) {
    case INTEGER:
        return p->value.i;
    case STRING:
    	printf("%s\n", p->value.s);
        break;
    case VARIABLE:
	IDfind(p->value.s, &val);
	return val;
    case WHILE:
	while (ex(p->SUB(0)))
	  ex(p->SUB(1));
	break;
    case IF:
	if (ex(p->SUB(0)))
	    ex(p->SUB(1));
	else if (p->value.sub.num > 2)
	    ex(p->SUB(2));
	break;
    case '=':
	name = p->SUB(0)->value.s;
        val = ex(p->SUB(1));
	if (IDfind(name, (int*)IDtest) == -1)
	  IDnew(INTEGER, name, val);
	else
	  IDreplace(INTEGER, name, val);
    	break;
    case READ:
	if (IDfind(p->value.s, 0) >= 0) {
	  if (scanf("%d", &val) == 1)
	    IDreplace(INTEGER, p->value.s, val);
	}
	break;
    case PRINT:  printf("%d\n", ex(p->SUB(0))); break;
    case ';':
	for (i = 0; i < p->value.sub.num; i++)
	  val = ex(p->SUB(i));
	return val;
    case UMINUS: return -ex(p->SUB(0));
    case '+':    return ex(p->SUB(0))  + ex(p->SUB(1));
    case '-':    return ex(p->SUB(0))  - ex(p->SUB(1));
    case '*':    return ex(p->SUB(0))  * ex(p->SUB(1));
    case '/':    return ex(p->SUB(0))  / ex(p->SUB(1));
    case '%':    return ex(p->SUB(0))  % ex(p->SUB(1));
    case '<':    return ex(p->SUB(0))  < ex(p->SUB(1));
    case '>':    return ex(p->SUB(0))  > ex(p->SUB(1));
    case GE:     return ex(p->SUB(0)) >= ex(p->SUB(1));
    case LE:     return ex(p->SUB(0)) <= ex(p->SUB(1));
    case NE:     return ex(p->SUB(0)) != ex(p->SUB(1));
    case EQ:     return ex(p->SUB(0)) == ex(p->SUB(1));
    default:     printf("unknown %d('%c')\n", p->attrib, p->attrib);
  }
  return 0;
}
